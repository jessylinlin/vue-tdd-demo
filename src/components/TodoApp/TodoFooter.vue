<template>
    <footer class="footer">
        <!-- This should be `0 items left` by default -->
        <span data-testid="todo-count" class="todo-count"><strong>{{ doneTodosCount }}</strong> item left</span>
        <!-- Remove this if you don't implement routing -->
        <ul class="filters">
            <li>
                <router-link exact to="/">All</router-link>
            </li>
            <li>
                <router-link to="/active">Active</router-link>
            </li>
            <li>
               <router-link to="/completed">Completed</router-link>
            </li>
        </ul>
        <!-- Hidden if no completed items are left â†“ -->
        <button 
            class="clear-completed" 
            data-testid="clear-completed" 
            v-if="isClearCompletedShow"
            @click="$emit('clear-completed')"
        >Clear completed</button>
    </footer>
</template>
<script>
export default {
    name: 'TodoFooter',
    props: {
        todos: {
            type: Array,
            required: true
        }
    },
    computed: {
        doneTodosCount() {
            return this.todos.filter(t => !t.done).length
        },
        isClearCompletedShow() {
            return this.todos.find(t => t.done)
        }
    }
}
</script>