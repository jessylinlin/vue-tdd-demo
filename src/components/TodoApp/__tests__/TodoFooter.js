import { createLocalVue, mount } from '@vue/test-utils'
import TodoFooter from '@/components/TodoApp/TodoFooter.vue'
import VueRouter from 'vue-router'

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter({
  linkActiveClass: 'selected'
})

describe('TodoFooter.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper */
  let wrapper = null

  beforeEach(async () => {
    const todos = [
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'sleep', done: true }
    ]
    wrapper = mount(TodoFooter, {
      propsData: {
        todos
      },
      localVue,
      router
    })
  })

  test('done todos count', () => {
    const count = wrapper.vm.todos.filter(t => !t.done).length
    const countEl = wrapper.find('[data-testid="todo-count"]')
    expect(Number.parseInt(countEl.text())).toBe(count)
  })
  test('clear completed show', async () => {
    const clearBtn = wrapper.find('[data-testid="clear-completed"]')
    expect(clearBtn.exists()).toBeTruthy()

    // 清除所有任务的完成状态
    wrapper.vm.todos.forEach(t => {
      t.done = false
    })

    wrapper = mount(TodoFooter, {
      propsData: {
        todos: [
          { id: 1, text: 'eat', done: false },
          { id: 2, text: 'sleep', done: false }
        ]
      },
      localVue,
      router
    })

    expect(wrapper.find('[data-testid="clear-completed"]').exists()).toBeFalsy()
  })
  test('clear completed', () => {
    const clearBtn = wrapper.find('[data-testid="clear-completed"]')
    // 断言是否对外发布自定义事件
    clearBtn.trigger('click')
    expect(wrapper.emitted()['clear-completed']).toBeTruthy()
  })

  test('router-link activeClass', async () => {
    const links = wrapper.findAllComponents({ name: 'RouterLink' })
    router.push('/active')

    await localVue.nextTick()

    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link.vm.to === '/active') {
        expect(link.classes()).toContain('selected')
      } else {
        expect(link.classes('selected')).toBeFalsy()
      }
    }
  })

  test('snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
