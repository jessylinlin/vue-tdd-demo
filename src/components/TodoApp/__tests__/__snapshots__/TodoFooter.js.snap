// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooter.vue snapshot 1`] = `
<footer class="footer"><span data-testid="todo-count" class="todo-count"><strong>1</strong> item left</span>
  <ul class="filters">
    <li><a href="#/" class="">All</a></li>
    <li><a href="#/active" aria-current="page" class="router-link-exact-active selected">Active</a></li>
    <li><a href="#/completed" class="">Completed</a></li>
  </ul> <button data-testid="clear-completed" class="clear-completed">Clear completed</button>
</footer>
`;
